### DISCUSSION TITLE
`or` condition?

### DISCUSSION THREAD

**Discussion by mwskwong** [open-circle/valibot#362] [Q&A] [answered]
I wish to make the `email` field:
  1. Optional, and either
  2. A literal `""`, or
  3. A valid email

For implementing (2) and (3), I used to use `z.or()` in Zod, is there an equivalent of that for Valibot?

**Comment by mwskwong** [comment]
OK, I guess I figered it out. `union` seems to do what I'm looking for.

```ts
const email = optional(
  union(
    [string([email()]), literal("")], 
    "Not a valid email"
  )
);
```

Note that, unlike Zod, the custom error message must associated with `union()` instead of `email()` or `literal()`.

**Reply by fabian-hiller** [reply]
Yes, that's how I would write the schema. Please let me know if you have any ideas on how to improve the API or DX for this use case. I know this is often used with forms. Maybe we can add a schema similar to [`optional`](https://valibot.dev/api/optional/) that adds this automatically. Read more here: https://valibot.dev/guides/optionals/

```ts
// Example that only works with `string`
const EmailSchema = maybeEmpty(string([email()]));

// Example that could work with any primitiv value
const EmailSchema = maybe(string([email()]), '');
```

**Reply by mwskwong** [reply]
@fabian-hiller Made a blog for my migration experience https://mwskwong.com/blog/migrating-from-zod-to-valibot-a-comparative-experience.

**TL;DR** I personally feel that the feature set of Valibot is comprehensive enough at this stage. What is lacking though, is the documentation. Currently, a lot of the pages are stubs and require some serious work on it. I'm happy to contribute if you agree with this direction.

**Reply by fabian-hiller** [reply]
Great article! I shared it on Twitter. I would love to get further feedback from you on our documentation! It would help us a lot to increase the adoption rate over the next months. Feel free to list every detail that is not perfect yet. You can create an issue for that or reach out to me on Discord (fabianhiller). We have also an issue tracking the progress of the API reference #287.

**Reply by fabian-hiller** [reply]
Thank you again for your feedback! It is very important for the further development of the library. As a first step, I started writing the Zod to Valibot migration guide: https://valibot.dev/guides/migrate-from-zod/

**Reply by mwskwong** [reply]
Thanks a lot.

IMO, the first thing that we should aim for is to complete the API reference. I will see what I can do at https://github.com/fabian-hiller/valibot/issues/287. But given the time constraint and workload, I would properly pick it up one by one by referencing existing PRs to give myself and others some flexibility.

**Reply by mwskwong** [reply]
Properly wanna pin that issue BTW. And update the title to invite contributors to contribute on the doc.

**Reply by fabian-hiller** [reply]
By the way... I think my post about your article is now one of my most popular posts on Twitter. So thanks a lot for the attention it created for the project. 🧩

**Reply by fabian-hiller** [reply]
> And update the title to invite contributors to contribute on the doc.

What would you call the title?

**Reply by mwskwong** [reply]
Properly something like `[Help Wanted][Documentation] Completing the API Reference`.

**Comment by fabian-hiller** [comment, marked as answer]
Note: With v0.26.0 the behavior has changed to improve the DX. The schema must now be written as follows:

```ts
import * as v from 'valibot';

const EmailSchema = v.optional(
  v.union([v.string([v.email('Not a valid email')]), v.literal('')])
);
```
